Bluetooth: Host: Optimize hci buf alloc by calc actually require size

LingaoM opened the pull request:
By calculating the actual space required instead of using the maximum space, 
for `ad_stream`, add two related APIs for save and restore status to calculate the total space.

Fixes #37710

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

Thalley commented:
I guess this fixes https://github.com/zephyrproject-rtos/zephyr/issues/37710

If we go down this road, should we not apply the same fix for periodic advertising as well?

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -524,27 +524,34 @@ static int hci_set_ad(uint16_t hci_op, const struct bt_ad *ad, size_t ad_len)
 		max_data_size = BT_GAP_ADV_MAX_ADV_DATA_LEN;
 	}
 
+	if (total_len > max_data_size) {
+		return -EINVAL;
+	}
```
Since we already check `if (total_len_bytes <= BT_HCI_LE_EXT_ADV_FRAG_MAX_LEN)` before calling this function, this check could be move into the non-ext `else` clause.

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -524,27 +524,34 @@ static int hci_set_ad(uint16_t hci_op, const struct bt_ad *ad, size_t ad_len)
+		return -EINVAL;
+	}
+
+	max_data_size = total_len;
```
Instead of overwriting `max_data_len`, could we set it to the "correct" value in the above `if` statement (`if (atomic_test_bit(adv->flags, BT_ADV_EXT_ADV))`) first time around? 

I think that's going to be more clear.

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -524,27 +524,34 @@ static int hci_set_ad(uint16_t hci_op, const struct bt_ad *ad, size_t ad_len)
+	}
+
+	max_data_size = total_len;
+	total_len += offsetof(struct bt_hci_cp_le_set_ext_adv_data, data);
```
I would actually prefer to create a new `const` variable which is something like
```c
const size_t cmd_size = total_len + offsetof(struct bt_hci_cp_le_set_ext_adv_data, data);
```

and use that instead of incrementing the `total_len`. `total_len` should then also become `total_data_len`. 

Alternative instead of using `offsetof` is to modify the struct from 
```c
struct bt_hci_cp_le_set_ext_adv_data {
	uint8_t  handle;
	uint8_t  op;
	uint8_t  frag_pref;
	uint8_t  len;
	uint8_t  data[251];
} __packed;
```
to
```c
struct bt_hci_cp_le_set_ext_adv_data {
	uint8_t  handle;
	uint8_t  op;
	uint8_t  frag_pref;
	uint8_t  len;
	uint8_t  data[0];
} __packed;
```
and then in all use cases when creating the cmd use something like
```c
bt_hci_cmd_create(BT_HCI_OP_LE_SET_EXT_ADV_DATA, sizeof(struct bt_hci_cp_le_set_ext_adv_data) + data_len; 
```

which we do for many other dynamic sized HCI commands.

LingaoM replied:
I prefer to use `offsetof`, because the latter needs to be changed at a larger point and has more uncertainty.

Thalley reviewed (changes requested):
It is a step in the right direction, but it feels half-way implemented. 

It handles only a specific case, but should probably be applied to periodic advertising as well, as well as optimizing when setting long data as well (e.g. if data is fragment in {50, 50, 5}, then the last will still send more bytes than needed).

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -572,17 +581,26 @@ static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op
 
-		buf = bt_hci_cmd_create(hci_op, sizeof(*set_data));
+		(void)memcpy(&backup, &stream, sizeof(stream));
+
+		data_len = ad_stream_read(&stream, NULL, sizeof(set_data->data));
+		cmd_size = data_len + offsetof(struct bt_hci_cp_le_set_ext_adv_data, data);
+
+		buf = bt_hci_cmd_create(hci_op, cmd_size);
 		if (!buf) {
 			return -ENOBUFS;
 		}
 
-		set_data = net_buf_add(buf, sizeof(*set_data));
+		set_data = net_buf_add(buf, cmd_size);
 
 		set_data->handle = adv->handle;
 		set_data->frag_pref = BT_HCI_LE_EXT_ADV_FRAG_ENABLED;
+
+		(void)memcpy(&stream, &backup, sizeof(stream));
+
```
Hmm, so you are pretty much using `ad_stream_read` to get the data length, before creating the buffer, but in order to do that you had to refactor `ad_stream_read` and then also create a backup for the stream. 

In that case, I think a new function to simply get the data_len without having to complicate the `ad_stream_read` function as well as avoiding creating a backup would be a simpler and more clean solution.

LingaoM replied:
I also prefer the view of @alwa-nordic , because the current implementation is still the simplest, and rewriting one is not necessarily better.

Thalley replied:
I like the idea that @alwa-nordic proposed, i.e. getting the total length in `ad_stream_new`. In that case, we should be able to avoid the save/restore functions, as well as the change to `ad_stream_read`, right?

LingaoM replied:
Please Review Again.

alwa-nordic commented on subsys/bluetooth/host/adv.c:
```diff
@@ -126,21 +126,30 @@ static uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t bu
 
 		if (stream->current_ltv_offset == 0) {
-			buf[read_len] = current_ltv->data_len + 1;
+			if (buf) {
```
These changes need to be reflected in the function documentation.

alwa-nordic commented on subsys/bluetooth/host/adv.c:
```diff
@@ -572,17 +581,26 @@ static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op
-		buf = bt_hci_cmd_create(hci_op, sizeof(*set_data));
+		(void)memcpy(&backup, &stream, sizeof(stream));
+
+		data_len = ad_stream_read(&stream, NULL, sizeof(set_data->data));
```
I prefer that the copy `backup` is what is used up. It's more function-like then.

I think the stream copy and read must be a new API method `ad_stream_peek`. The fact `ad_stream` can be shallow-copied should be an implementation detail.

alwa-nordic reviewed (changes requested):
The commit message also needs work. I suggest changing the subject line to "Optimize hci buf allocation by calculating `ad_stream` size". The commit message body could use some love as well :). Please also add an abstract of the changes / solution.

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -524,29 +535,29 @@ static int hci_set_ad(uint16_t hci_op, const struct bt_ad *ad, size_t ad_len)
-
-	err = set_data_add_complete(set_data->data, max_data_size, ad, ad_len,
-				    &set_data->len);
+	err = set_data_add_complete(set_data->data, total_data_len,
+				    ad, ad_len, &set_data->len);
```
Correct me if I am wrong, but won't `total_data_len` and `set_data->len` always be the same value after this call?

LingaoM replied:
Yes, always same, so we can use `(void)set_data_add_complete`.

Good point.

Thalley replied:
> Yes, always same, so we can use `(void)set_data_add_complete`.
> 
> Good point.

How did you come to that conclusion? 

It does actually looks like that in the case of `BT_DATA_NAME_COMPLETE` that is shortened that they will not be equal. In that case, the calculated length is larger than &set_data->len. 

The case where shortening the name fails due to the size can still happen here, can't it?

LingaoM replied:
`if ((set_data_len + len + 2) > set_data_len_max) `

This judgment will also false, because we do same calculation for set_data_len_max before. Only problem is we can't change type to short_name，or we not need todo this, i think.

LingaoM replied:
``` C
	/* Provide the opportunity to truncate the complete name */
	if (!atomic_test_bit(adv->flags, BT_ADV_EXT_ADV) &&
	    total_data_len > BT_GAP_ADV_MAX_ADV_DATA_LEN) {
		total_data_len = BT_GAP_ADV_MAX_ADV_DATA_LEN;
	}
```

Solved.

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -51,11 +51,21 @@ struct ad_stream {
+
+	for (int i = 0; i < ad_len; i++) {
+		for (int j = 0; j < ad[i].len; j++) {
+			len += ad[i].data[j].data_len + 2;
```
While it may be obvious to us (and right now), please add a comment stating that `+ 2` is from the LTV header (length and type).

alwa-nordic commented on subsys/bluetooth/host/adv.c:
```diff
@@ -629,7 +640,7 @@ static int hci_set_ad_ext(struct bt_le_ext_adv *adv, uint16_t hci_op,
 		 * This allows us to update advertising data while advertising.
 		 */
-		return hci_set_adv_ext_complete(adv, hci_op, ad, ad_len);
+		return hci_set_adv_ext_complete(adv, hci_op, total_len_bytes, ad, ad_len);
```
This change is independent and should go in a separate commit. Maybe even a separate PR. It's easier for us reviewers when the PRs are kept small. (I know @Thalley asked for this in this PR, but personally I would prefer to review one change at a time.)

LingaoM replied:
I recommend using a separate commit instead of a separate pr.

LingaoM committed:
Bluetooth: Host: Optimize hci buf alloc by calc actually require size

By calculating the actual space required instead of using
the maximum space, for `hci_set_adv_ext_complete`.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/37710

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Host: Reconstruct stream interface

Add `remaining_size` in struct ad_stream.

By calculating the actual space required instead of
using the maximum space.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented on subsys/bluetooth/host/adv.c:
```diff
@@ -524,34 +535,29 @@ static int hci_set_ad(uint16_t hci_op, const struct bt_ad *ad, size_t ad_len)
-	if (!buf) {
-		return -ENOBUFS;
+	if (!atomic_test_bit(adv->flags, BT_ADV_EXT_ADV) &&
+	    total_data_len > BT_GAP_ADV_MAX_ADV_DATA_LEN) {
```
@Thalley and we already check this before.

So the length of advertising always can place and not need to shorten.

LingaoM commented on subsys/bluetooth/host/adv.c:
```diff
@@ -49,13 +49,23 @@ struct ad_stream {
+	for (int i = 0; i < ad_len; i++) {
+		for (int j = 0; j < ad[i].len; j++) {
+			/* LTV length + type + value */
+			stream->remaining_size += ad[i].data[j].data_len + 2;
```
remaining_size `uint16_t` max 65535, how can overflow?

Thalley replied:
since ad_len is a `size_t`, the application can technically attempt to set 22000 entries in the array, and if they are each 3 octets (length + type + 1 byte data) that will be a total of 66000 octets here, and that would cause an overflow, and `stream->remaining_size` will be `465`.

LingaoM replied:
Done

Thalley replied:
What is "done" here? I don't see any change

LingaoM replied:
https://github.com/zephyrproject-rtos/zephyr/pull/43255/files#r822427197

Thalley replied:
I'm sorry, I don't fully understand your comment. 

My point is that you are here, without conditions, increasing a `size_t` value by looping over two other `size_t` values. Without any checks for `ad_len` or `ad[i].len` this can cause overflow.

LingaoM replied:
```suggestion
			please provide your modification
```

LingaoM replied:
Now that the type of remaining_size  is already size_t, how can this possibly overflow?

LingaoM replied:
According to your opinion, `current_ltv_offset` this parameter may still overflow, isn't it?

LingaoM replied:
If four bytes of space cannot represent the length of the total data, the use of the application layer is undoubtedly stupid.

Thalley replied:
> If four bytes of space cannot represent the length of the total data, the use of the application layer is undoubtedly stupid.

Right, and we should protect the stack against a buggy (or stupid) application. 

We know that the maximum advertising data can be 1650. 

I would suggest simply adding
```c
if (stream->remaining_size > BT_GAP_ADV_MAX_EXT_ADV_DATA_LEN) {
	/* return err */
}
```

Which not only guards against overflow, but also attempted spec violation.

LingaoM commented on subsys/bluetooth/host/adv.c:
```diff
@@ -146,6 +155,8 @@ static uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t bu
 		}
 	}
 
+	stream->remaining_size -= read_len;
```
`stream->remaining_size` represents the sum of all lengths, which is always greater than read_ Len, isn't it? I can't see it underflow

Thalley replied:
It should be indeed :) Unless we have an error somewhere. IMO it doesn't hurt to add an _ASSERT to verify that `stream->remaining-size > read_len`

LingaoM replied:
Done :)

Thalley commented:
> I guess this fixes #37710
> 
> If we go down this road, should we not apply the same fix for periodic advertising as well?

@LingaoM Can you please add the `fixes` keyword for the PR description to link it to this issue?

LingaoM commented:
@Thalley @alwa-nordic @jhedberg If there is no problem, please approve

LingaoM commented on subsys/bluetooth/host/adv.c:
```diff
@@ -49,13 +49,23 @@ struct ad_stream {
 	uint16_t current_ltv_offset;
+
+	/* The remaining size of total ad[i].data[j].data_len + 2 for LTV header */
+	size_t remaining_size;
```
https://github.com/zephyrproject-rtos/zephyr/pull/43255/files#r817569322

Not and modified?
